drivers/mtd_flashpage/mtd_flashpage.c

```diff
@@ -40,9 +45,17 @@ static int _init(mtd_dev_t *dev)
 
 static int _read(mtd_dev_t *dev, void *buf, uint32_t addr, uint32_t size)
 {
-    assert(addr < MTD_FLASHPAGE_END_ADDR);
+    mtd_flashpage_t *super = container_of(dev, mtd_flashpage_t, base);
 
-    (void)dev;
+    /* we need to accept both aliased and non-aliased flash addresses */
+    if (addr > CPU_FLASH_BASE) {
+        addr += (uintptr_t)flashpage_addr(super->offset) - CPU_FLASH_BASE;
+    } else {
+        addr += (uintptr_t)flashpage_addr(super->offset);
+    }
```

Comment:
What is the reason for this? It might be risky if the flash is larger than `CPU_FLASH_BASE`.

Reply:
That's mostly there to make `tests/mtd_flashpage` (and any existing users if there are any) happy.

That test operates on raw flash addresses obtained from `flashpage_addr()` and expects to use those on the MTD device.

On a MTD device however addresses are relative to the start of the memory, not some absolute physical address.

So when we have 64k of Flash at `0x08000000` that test will try to access the last byte as `0x0800FFFF` whereas any 'sane' MTD user that doesn't know about the underlying hardware would just access `0xFFFF`.

Now we could just as well get rid of that leaky abstraction and fix the test, but then I thought the test might be there for a reason and something actually relies on this wacky behavior.

Reply:
I really would opt for fixing the test than bringing such behavior into the actual driver. It should be consistent.

Reply:
I would prefer that too, but I'm afraid of breaking existing users.
In-tree we only have `tests/mtd_flashpage` and `sys/fido2` - the latter should really use the AUX page as it's current use of raw flashpage addresses is very unsafe.

We should definitely deprecate the old mtd_flashpage access style as it is just based on the hope that no important data is at the defined address without any way of actually guaranteeing that.

Reply:
Ah nvm, the solution is actually [really simple](https://github.com/RIOT-OS/RIOT/pull/18608/commits/e95ac24b17a68db94a9e899efc13b5d078d0d1c6)